src/core/ext/transport/chttp2/transport/hpack_table.h

```diff
@@ -46,41 +46,65 @@
 #endif
 
 /* hpack decoder table */
-typedef struct {
+struct grpc_chttp2_hptbl {
+  static uint32_t entries_for_bytes(uint32_t bytes) {
+    return (bytes + GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD - 1) /
+           GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD;
+  }
+  static constexpr uint32_t InitialCapacity =
+      (/*bytes=*/GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE +
+       GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD - 1) /
+      GRPC_CHTTP2_HPACK_ENTRY_OVERHEAD;
+
+  grpc_chttp2_hptbl() = default;
+
   /* the first used entry in ents */
-  uint32_t first_ent;
+  uint32_t first_ent = 0;
   /* how many entries are in the table */
-  uint32_t num_ents;
+  uint32_t num_ents = 0;
   /* the amount of memory used by the table, according to the hpack algorithm */
-  uint32_t mem_used;
+  uint32_t mem_used = 0;
   /* the max memory allowed to be used by the table, according to the hpack
      algorithm */
-  uint32_t max_bytes;
+  uint32_t max_bytes = GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE;
   /* the currently agreed size of the table, according to the hpack algorithm */
-  uint32_t current_table_bytes;
+  uint32_t current_table_bytes = GRPC_CHTTP2_INITIAL_HPACK_TABLE_SIZE;
   /* Maximum number of entries we could possibly fit in the table, given defined
      overheads */
-  uint32_t max_entries;
+  uint32_t max_entries = InitialCapacity;
   /* Number of entries allocated in ents */
-  uint32_t cap_entries;
+  uint32_t cap_entries = InitialCapacity;
   /* a circular buffer of headers - this is stored in the opposite order to
      what hpack specifies, in order to simplify table management a little...
      meaning lookups need to SUBTRACT from the end position */
-  grpc_mdelem* ents;
-  grpc_mdelem static_ents[GRPC_CHTTP2_LAST_STATIC_ENTRY];
-} grpc_chttp2_hptbl;
+  grpc_mdelem* ents = nullptr;
+};
 
 /* initialize a hpack table */
-void grpc_chttp2_hptbl_init(grpc_chttp2_hptbl* tbl);
+inline void grpc_chttp2_hptbl_init(grpc_chttp2_hptbl* tbl) {
+  new (tbl) grpc_chttp2_hptbl();
+  constexpr uint32_t AllocSize =
+      sizeof(*tbl->ents) * grpc_chttp2_hptbl::InitialCapacity;
+  tbl->ents = static_cast<grpc_mdelem*>(gpr_malloc(AllocSize));
+  memset(tbl->ents, 0, AllocSize);
```

Comment:
Please move line 86-89 into the ctor.

Reply:
That causes ASAN to get mad by causing a memory leak.

The issue: hpack_table is embedded inside the CH2 parser, which uses grpc_core::New.
So if this malloc/memset is in the constructor:
1) The transport is created, which calls new() for CH2, which has hpack_table embedded, so it calls new.
2) Later on hpack_parser_init is called, which calls hpack_table_init, which in my original PR also called new.

This causes a leak.

It's possible, perhaps, to:
1) Move it into the constructor as you said
2) Have init() call the constructor
3) Remove table_init() from parser_init()

WDYT?

Reply:
So, for my understanding: you're basically saying the ctor is called twice, right?  If that's the case, we should restructure the code that it doesn't happen (perhaps that's what you mean by 3?).

If that's not the case, I'll need a little bit of help to understand what's going on.

Reply:
Moved it into the constructor. Changed the type from a raw pointer to a unique array (added template specialization to src/core/lib/gprpp/memory.h). That way, even if we have multiple inits (e.g. unit tests) we do the right thing.